sources/web/datalab/jupyter.ts

```diff
@@ -81,67 +88,96 @@ function pipeOutput(stream: NodeJS.ReadableStream, port: number, error: boolean)
-    target: 'http://127.0.0.1:' + server.port
-  };
-
-  server.proxy = httpProxy.createProxyServer(proxyOptions);
-  server.proxy.on('proxyRes', responseHandler);
-  server.proxy.on('error', errorHandler);
-
-  return server;
+  tcp.waitUntilFree(port).then(
+    function() {
+      var userDir = userManager.getUserDir(userId);
+      if (!fs.existsSync(userDir)) {
+        fs.mkdirSync(userDir, parseInt('0755', 8));
+      }
+
+      var server: JupyterServer = {
+        userId: userId,
+        port: port,
+        notebooks: userDir,
+        portResolved: function() {},
+      };
+
+      function exitHandler(code: number, signal: string): void {
+        logging.getLogger().error('Jupyter process %d for user %s exited due to signal: %s',
+                                  server.childProcess.pid, userId, signal);
+        delete jupyterServers[server.userId];
+      }
+
+      var processArgs = appSettings.jupyterArgs.slice().concat([
+        '--port=' + server.port,
+        '--port-retries=0',
+        '--notebook-dir="' + server.notebooks + '"'
+      ]);
+
+      var notebookEnv: any = process.env;
+      if ('KG_URL' in notebookEnv && notebookEnv['KG_URL']) {
+        logging.getLogger().info(
+          'Found a kernel gateway URL of %s... configuring the notebook to use it', notebookEnv['KG_URL']);
+        processArgs = processArgs.concat([
+          '--NotebookApp.session_manager_class=nb2kg.managers.SessionManager',
+          '--NotebookApp.kernel_manager_class=nb2kg.managers.RemoteKernelManager',
+          '--NotebookApp.kernel_spec_manager_class=nb2kg.managers.RemoteKernelSpecManager'
+        ]);
+      }
+
+      var processOptions = {
+        detached: false,
+        env: notebookEnv
+      };
+
+      server.childProcess = childProcess.spawn('jupyter', processArgs, processOptions);
+      server.childProcess.on('exit', exitHandler);
+      logging.getLogger().info('Jupyter process for user %s started with pid %d and args %j',
+                               userId, server.childProcess.pid, processArgs);
+
+      server.portResolved = function() {
+        // Ensure that the portResolved function only gets invoked once.
+        server.portResolved = function(){};
+
+        // Create the proxy.
+        var proxyOptions: httpProxy.ProxyServerOptions = {
+          target: 'http://127.0.0.1:' + port
+        };
+
+        server.proxy = httpProxy.createProxyServer(proxyOptions);
+        server.proxy.on('proxyRes', responseHandler);
+        server.proxy.on('error', errorHandler);
+
+        tcp.waitUntilUsed(server.port, 100, 3000).then(
+          function() {
+            jupyterServers[userId] = server;
+            logging.getLogger().info('Jupyter server started for %s.', userId);
+            callbackManager.invokeAllCallbacks(userId, null);
+          },
+          function(e) {
+            logging.getLogger().error(e, 'Failed to start Jupyter server for user %s.', userId);
```

Comment:
> we don't retry on failures here?

No, not once we get to this point. We're only retrying failures in picking an unused port; once we've picked an unused port, any failures after that are just cascaded upward.